internal/utils/gke.go

```diff
@@ -0,0 +1,486 @@
+	gkeapi "google.golang.org/api/container/v1"
+	"google.golang.org/api/option"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// Node Pool Status
+const (
+	// NodePoolStatusUnspecified - Not set.
+	NodePoolStatusUnspecified = "STATUS_UNSPECIFIED"
+	// NodePoolStatusProvisioning The PROVISIONING state indicates the node pool is
+	// being created.
+	NodePoolStatusProvisioning = "PROVISIONING"
+	// NodePoolStatusRunning The RUNNING state indicates the node pool has been
+	// created
+	// and is fully usable.
+	NodePoolStatusRunning = "RUNNING"
+	// NodePoolStatusRunningWithError The RUNNING_WITH_ERROR state indicates the
+	// node pool has been created
+	// and is partially usable. Some error state has occurred and
+	// some
+	// functionality may be impaired. Customer may need to reissue a
+	// request
+	// or trigger a new update.
+	NodePoolStatusRunningWithError = "RUNNING_WITH_ERROR"
+	// NodePoolStatusReconciling The RECONCILING state indicates that some work is
+	// actively being done on
+	// the node pool, such as upgrading node software. Details can
+	// be found in the `statusMessage` field.
+	NodePoolStatusReconciling = "RECONCILING"
+	//  NodePoolStatusStopping The STOPPING state indicates the node pool is being
+	// deleted.
+	NodePoolStatusStopping = "STOPPING"
+	// NodePoolStatusError The ERROR state indicates the node pool may be unusable.
+	// Details
+	NodePoolStatusError = "ERROR"
+)
+
+// Network Providers
+const (
+	// NetworkProviderUnspecified is an unspecified netowrk provider
+	NetworkProviderUnspecified = "PROVIDER_UNSPECIFIED"
+
+	// NetworkProviderCalico describes the calico provider
+	NetworkProviderCalico = "CALICO"
+)
+
+// Cluster Status
+const (
+	// ClusterStatusProvisioning The PROVISIONING state indicates the cluster is
+	// being created.
+	ClusterStatusProvisioning = "PROVISIONING"
+
+	// ClusterStatusRunning The RUNNING state indicates the cluster has been
+	// created and is fully
+	ClusterStatusRunning = "RUNNING"
+
+	// ClusterStatusStopping The STOPPING state indicates the cluster is being
+	// deleted.
+	ClusterStatusStopping = "STOPPING"
+
+	// ClusterStatusError is a ClusterStatus enum value
+	ClusterStatusError = "ERROR"
+
+	// ClusterStatusReconciling The RECONCILING state indicates that some work is
+	// actively being done on
+	ClusterStatusReconciling = "RECONCILING"
+
+	// ClusterStatusUnspecified - Not set.
+	ClusterStatusUnspecified = "STATUS_UNSPECIFIED"
+
+	// ClusterStatusDegraded The DEGRADED state indicates the cluster requires user
+	// action to restore
+	ClusterStatusDegraded = "DEGRADED"
+)
+
+// Errors
+const (
+	cannotBeNilError            = "field [%s] cannot be nil for non-import cluster [%s]"
+	cannotBeNilForNodePoolError = "field [%s] cannot be nil for nodepool [%s] in non-nil cluster [%s]"
+)
+
+// GenerateGkeClusterCreateRequest creates a request
+func GenerateGkeClusterCreateRequest(config *gkev1.GKEClusterConfig) (*gkeapi.CreateClusterRequest, error) {
```

Comment:
Curious why you put this type of function in the utils and not the controller. Seems like its not used anywhere else.

Reply:
Ok we can move in the other PR.